adr/0003-add-facts-gathering-capabilities-to-the-agent.md

```diff
@@ -0,0 +1,28 @@
+# 3. Add facts-gathering capabilities to the agent
+
+Date: 2022-11-17
+
+## Status
+
+Accepted
+
+## Context
+
+We need to gather facts from the hosts to be able to use them in the expectations section of the checks.
+A "gatherer" is a piece of code that knows how to extract a specific fact from a host.
+We need to implement a way to run the gatherers and collect the facts, namely the "gathering engine".
+Also, advanced users might want to declare their gatherers to be used in the expectations section,
+we can refer to this as "gatherers plugins".
+
+## Decision
+
+The agent should be able to gather facts from the hosts and send them to the server.
+A request is sent from the server to the agent to gather facts, the agent runs the gatherers and sends the facts to the server.
+The communication between the agent and the server is asynchronously done via AMQP.
+If the agent is not able to gather one or more facts, it should send a message to the server with the list of the failed gatherers alongside the successfully gathered facts, in a best-effort fashion.
```

Comment:
I would add a last sentence saying that this allows the agent to query only once the "source of truth" of specific gatherers. For example: one query once the cibadmin command for each fact gathering process.
This helps to reduce any potential overload in the agents.
In the past, with ansible, without having this "cache" capability, we were running the commands for each particular check.

Reply:
nice i will add it!